app/renderer/utils/locator-generation.js

```diff
@@ -391,38 +415,126 @@ export function getOptimalPredicateString(doc, domNode) {
       }
 
-      // If the attribute isn't actually unique, get it's index too
+      // Return as soon as the accumulated attribute combination is unique
       if (othersWithAttr.length === 1) {
         return predicateString.join(' AND ');
       }
     }
   } catch (error) {
-    // If there's an unexpected exception, abort and don't get an XPath
-    log.error(
-      `The most optimal '-ios predicate string' could not be determined ` +
-        `because an error was thrown: '${JSON.stringify(error, null, 2)}'`,
-    );
+    // If there's an unexpected exception, abort
+    logLocatorError('predicate string', error);
+    return null;
+  }
+}
+
+/**
+ * Get an optimal UiAutomator selector for a Node
+ * Only works for elements inside the last direct child of the hierarchy (xpath: /hierarchy/*[last()] )
+ *
+ * @param {Document} doc
+ * @param {Node} domNode
+ * @param {string} path a dot-separated string of indices
+ * @returns {string|null}
+ */
+export function getOptimalUiAutomatorSelector(doc, domNode, path) {
+  try {
+    // BASE CASE #1: If this isn't an element, or we're above the root, return empty string
+    if (!domNode.tagName || domNode.nodeType !== 1) {
+      return '';
+    }
+
+    // UiAutomator can only find elements inside the last direct child of the hierarchy.
+    // hierarchy is the child of doc (which is <xml/>), so need to get the children of its child
+    // BASE CASE #2: If there is no hierarchy or its children, return null
+    const docChildren = childNodesOf(doc);
+    const hierarchyChildren = _.isEmpty(docChildren) ? [] : childNodesOf(docChildren[0]);
+    if (_.isEmpty(hierarchyChildren)) {
+      return null;
+    }
+
+    // BASE CASE #3: If looking for an element that is not inside
+    // the last direct child of the hierarchy, return null
+    const lastHierarchyChildIndex = (hierarchyChildren.length - 1).toString();
+    let pathArray = path.split('.');
+    const requestedHierarchyChildIndex = pathArray[0];
+    if (requestedHierarchyChildIndex !== lastHierarchyChildIndex) {
+      return null;
+    }
+
+    // In order to use only the last direct child of the hierarchy as the new scope,
+    // need to recreate it as a Document (Node -> XML -> Document),
+    // then modify the path by changing the first index,
+    // and finally recreate the domNode, since it still references the original parent
+    const lastHierarchyChild = hierarchyChildren[lastHierarchyChildIndex];
+    const newXml = xmlSerializer.serializeToString(lastHierarchyChild);
+    // wrap the new XML in a dummy tag which will have the node type Document
+    const newDoc = domParser.parseFromString(`<dummy>${newXml}</dummy>`);
+    pathArray[0] = '0';
+    const newPath = pathArray.join('.');
+    const newDomNode = findDOMNodeByPath(newPath, newDoc);
+
+    // BASE CASE #4: Check all attributes and try to find unique ones
+    let uiSelector, othersWithAttr, mostUniqueSelector;
+    let othersWithAttrMinCount = 99;
```

Comment:
what is this magic number?

Reply:
I guess `Number.MAX_SAFE_INTEGER` would make the intent more obvious. Another approach would be to skip the initial value here, and assign it in the following comparison if it is undefined.